### PR TITLE
Load templates from external repository

### PR DESCRIPTION
**Pull request by ubergesundheit**
PR Information is [here](https://github.com/sensebox/sensebox-mailer/pull/14#issuecomment-766969458)

### CODE REVIEW

**Comment by mpfeil** [comment]
Still WIP. Please do not change it. Going to document some more stuff on the weekend.

**Comment by mpfeil** [comment]
Finally @ubergesundheit and I finished our hackathon project. We removed the mail templates from this repository and moved them over to [sensebox-mailer-templates](https://github.com/sensebox/sensebox-mailer-templates).

The big change is: the `sensebox-mailer` (this repo) pulls the templates from the `sensebox-mailer-templates` repository.
So the `sensebox-mailer` must not be rebuild if you change a mail template or found a typo in one of the mail templates 🎉
By default the `sensebox-mailer` checks the templates repos every 5 minutes (the interval is configurable) for new templates and pulls them in.

Also this PR introduces Github Actions for CI.

File: README.md [hunk 1]
```diff
@@ -97,7 +97,7 @@ You should configure the following variables:
...
 | `SMTP_PASSWORD` | the smtp server password |  |
 | `FROM_DOMAIN` | the domain you are sending from |  |
-| `TEMPLATES_REPOSITORY` | |  |
-| `TEMPLATES_BRANCH` | |  |
-| `TEMPLATES_FS_PATH` | |  |
-| `TEMPLATES_FETCH_INTERVAL` | |  |
+| `TEMPLATES_REPOSITORY` | repository of the templates |  |
+| `TEMPLATES_BRANCH` | main branch of the repository which is pulled |  |
+| `TEMPLATES_FS_PATH` | path where to clone repository |  |
+| `TEMPLATES_FETCH_INTERVAL` | time interval how often to pull repository |  |
```
Comment (ubergesundheit) [inline README.md]: Aren't all of these optional?

```suggestion
| `TEMPLATES_REPOSITORY` | git url of the repository which contains the templates | y |
| `TEMPLATES_BRANCH` | use this branch of the repository for templates | y |
| `TEMPLATES_FS_PATH` | path where to clone repository | y |
| `TEMPLATES_FETCH_INTERVAL` | time interval how often to pull repository. Specify in [go duration](https://golang.org/pkg/time/#ParseDuration) format.  | y |
```

**Review by felixerdy — APPROVED** [review]
Looks good. Great work 📯

**Review by felixerdy — CHANGES_REQUESTED** [review]
I just deployed the new mailer on our testing system. Unfortunately, it's not working:
```
mailer_1_8e829b25ecdb | sensebox-mailer startup. Version: ci-stuff 2021-01-25T18:22:18Z a26d8c8
mailer_1_8e829b25ecdb | Templates do not exists; Go and clone repository
mailer_1_8e829b25ecdb | Cloning into './mailer-templates'...
mailer_1_8e829b25ecdb | Executing [git clone -b language https://github.com/sensebox/sensebox-mailer-templates.git ./mailer-templates]2021-02-22T08:38:35Z [StartHTTPSServer] senseBox Mailer startup
mailer_1_8e829b25ecdb | 2021-02-22T08:38:35Z [StartHTTPSServer] created client cert pool
mailer_1_8e829b25ecdb | 2021-02-22T08:38:35Z [StartHTTPSServer] imported server cert
mailer_1_8e829b25ecdb | 2021-02-22T08:38:35Z [StartHTTPSServer] built name to certificate
mailer_1_8e829b25ecdb | 2021-02-22T08:38:35Z [StartHTTPSServer] configured server
mailer_1_8e829b25ecdb | 2021-02-22T08:38:35Z [StartHTTPSServer] starting server on address 0.0.0.0:3924
mailer_1_8e829b25ecdb | 2021-02-22T08:38:46Z [requestHandler] incoming request
mailer_1_8e829b25ecdb | 2021-02-22T08:38:46Z [requestHandler] Error decoding JSON payload: Template not available (template: 'passwordReset', language: 'de_DE')
mailer_1_8e829b25ecdb | 2021-02-22T08:38:46Z [ServeHTTP] Error: Template not available (template: 'passwordReset', language: 'de_DE')
```

I already changed the language codes from 'de' to 'de_DE' in https://github.com/sensebox/sensebox-mailer-templates/tree/language but still it's not working. Any ideas?

File: mailer/templates/templates.go [hunk 2]
```diff
@@ -1,124 +1,129 @@
...
-	if err != nil {
-		return err
+func (templater *Templater) FetchLatestTemplatesFromGithub() {
+	ticker := time.NewTicker(templater.RepositoryPullInterval)
+
+	for range ticker.C {
+		cmd := exec.Command("git", "pull", "origin", templater.RepositoryBranch)
+		cmd.Dir = templater.RepositoryFsPath
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v", cmd.Args)
```
Comment (ubergesundheit) [inline mailer/templates/templates.go]:
```suggestion
		fmt.Printf("Executing %v\n", cmd.Args)
```

File: mailer/templates/templates.go [hunk 3]
```diff
@@ -1,124 +1,129 @@
...
+		err := cmd.Run()
+		if err != nil {
+			fmt.Printf("Error pulling git repository %v", err)
```
Comment (ubergesundheit) [inline mailer/templates/templates.go]:
```suggestion
			fmt.Printf("Error pulling git repository %v\n", err)
```

File: mailer/templates/templates.go [hunk 4]
```diff
@@ -1,124 +1,129 @@
...
+		}
+		err = templater.slurpTemplates(templater.RepositoryFsPath)
+		if err != nil {
+			fmt.Printf("Error reading templates %v", err)
```
Comment (ubergesundheit) [inline mailer/templates/templates.go]:
```suggestion
			fmt.Printf("Error reading templates %v\n", err)
```

File: mailer/templates/templates.go [hunk 5]
```diff
@@ -1,124 +1,129 @@
...
-func FromJSON() (int, error) {
-	if err := initStatikFS(); err != nil {
-		return 0, err
+func (templater *Templater) CloneTemplatesFromGitHub() error {
+	// Check if templates folder exists
+	if _, err := os.Stat(templater.RepositoryFsPath); os.IsNotExist(err) {
+		fmt.Println("Templates do not exists; Go and clone repository")
+		cmd := exec.Command("git", "clone", "-b", templater.RepositoryBranch, templater.RepositoryGitURL, templater.RepositoryFsPath)
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v", cmd.Args)
```
Comment (ubergesundheit) [inline mailer/templates/templates.go]:
```suggestion
		fmt.Printf("Executing %v\n", cmd.Args)
```

File: mailer/templates/templates.go [hunk 6]
```diff
@@ -1,124 +1,129 @@
...
+	})
 
-	err = json.Unmarshal(jsonBytes, &theTemplates)
 	if err != nil {
-		return 0, err
+		return err
 	}
 
-	return len(theTemplates), nil
+	templater.TheTemplates = templates
```
Comment (ubergesundheit) [inline mailer/templates/templates.go:114]:
```suggestion
	if len(templates) == 0 {
		return fmt.Errorf("No templates imported from template repository")
	}

	templater.TheTemplates = templates
```

**Comment by ubergesundheit** [comment]
I've added some linebreaks and a check for zero templates, maybe you could try it again with commit a98b5ed2345399fe4d5f4af8a947ed17b4e09128 ?

Otherwise I have no idea. I can check this evening (or probably later)

**Comment by ubergesundheit** [comment]
Oh btw did you change the `TEMPLATES_BRANCH` to test your changes in the templates repository? But IMO this `de`/`de_DE` change is not the issue (because this is handleed in code [here](https://github.com/sensebox/sensebox-mailer/pull/14/files#diff-dfe82bb7b125bf34007347d94bb3e0015151c6eafb07b1ee143d02dd6dfc24e9R127))

**Comment by felixerdy** [comment]
Still not working, resulting in the same error like above. Yes, I used the `TEMPLATES_BRANCH` and also tested again with the default `main` branch but no avail.

File: mailer/templates/templates.go [hunk 7]
```diff
@@ -1,124 +1,133 @@
...
+func (templater *Templater) CloneTemplatesFromGitHub() error {
+	// Check if templates folder exists
+	if _, err := os.Stat(templater.RepositoryFsPath); os.IsNotExist(err) {
+		fmt.Println("Templates do not exists; Go and clone repository")
+		cmd := exec.Command("git", "clone", "-b", templater.RepositoryBranch, templater.RepositoryGitURL, templater.RepositoryFsPath)
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v\n", cmd.Args)
+		return cmd.Run()
 	}
+	return templater.slurpTemplates(templater.RepositoryFsPath)
```
Comment (felixerdy) [inline mailer/templates/templates.go]: I just realized it's working, but only after 5 min when it runs `git pull ...` the first time. So in `CloneTemplatesFromGitHub`, is it executing `templater.slurpTemplates(templater.RepositoryFsPath)` even though it does `return cmd.Run()` beforehand?

Comment (ubergesundheit) [reply mailer/templates/templates.go]: I think this should be correct, the if block checks if the directory does exist, it clones the repository.

But the observation is really good. Could reproduce that on my machine! I'll investigate later

File: mailer/templates/templates.go [hunk 8]
```diff
@@ -1,124 +1,133 @@
...
-	if err := initStatikFS(); err != nil {
-		return 0, err
+func (templater *Templater) CloneTemplatesFromGitHub() error {
+	// Check if templates folder exists
+	if _, err := os.Stat(templater.RepositoryFsPath); os.IsNotExist(err) {
+		fmt.Println("Templates do not exists; Go and clone repository")
+		cmd := exec.Command("git", "clone", "-b", templater.RepositoryBranch, templater.RepositoryGitURL, templater.RepositoryFsPath)
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v\n", cmd.Args)
+		return cmd.Run()
```
Comment (ubergesundheit) [inline mailer/templates/templates.go]:
```suggestion
		err := cmd.Run()
		if err != nil {
			return err
		}
```

File: mailer/templates/templates.go (hunk 7 above)
Comment (ubergesundheit) [reply mailer/templates/templates.go]: https://github.com/sensebox/sensebox-mailer/pull/14/commits/ea11cf19747e71d172155746743ca619d6e5fc66 could fix that

**Review by felixerdy — APPROVED** [review]

**Comment by felixerdy** [comment]
Great, it's working now!

**Pull request merged**